docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

```diff
@@ -0,0 +1,213 @@
+- `gardenctlv2` (or any other client)
+  - Creates `Bastion` resource in the garden cluster
+  - Establishes an `ssh` connection to a shoot node, using a bastion host as proxy
+  - Heartbeats / keeps alive the `Bastion` resource during `ssh` connection
+- Gardener extension provider <infra>
+  - Provider specific bastion controller
+  - Should be added to gardener-extension-provider-<infra> repos, e.g. https://github.com/gardener/gardener-extension-provider-aws/tree/master/pkg/controller
+  - Has the permission to update the `Bastion/status` subresource on the seed cluster
+  - Runs on seed (of course)
+- Gardener Controller Manager (`GCM`)
+  - `Bastion` heartbeat controller
+    - Cleans up `Bastion` resource on missing heartbeat.
+    - Is configured with a `maxLifetime` and `timeToLife` for the `Bastion` resource
+- Gardener (RBAC)
+  - The project `admin` role should be extended to allow CRUD operations on the `Bastion` resource. The `gardener.cloud:system:project-member-aggregation` `ClusterRole` needs to be updated accordingly (https://github.com/gardener/gardener/blob/master/charts/gardener/controlplane/charts/application/templates/rbac-user.yaml)
+
+### SSH Flow
+0. Users should only get the RBAC permission to `create` / `update` `Bastion` resources for a namespace, if they should be allowed to `ssh` onto the shoot nodes in this namespace. A project member with `admin` role will have these permissions.
+1. User/`gardenctlv2` creates `Bastion` resource in garden cluster (see [resource example](#resource-example) below)
+    - First, gardenctl would figure out the own public IP of the user's machine. Either by calling an external service (gardenctl (v1) uses https://github.com/gardener/gardenctl/blob/master/pkg/cmd/miscellaneous.go#L226) or by calling a binary that prints the public IP(s) to stdout. The binary should be configurable. The result is set under `spec.ingress[].from[].ipBlock.cidr`
+    - the public `ssh` key of the user is set under `spec.sshPublicKey`. The key that should be used needs to be configured beforehand by the user
+    - The targeted shoot is set under `spec.shootRef`
+2. GAPI Admission Control for the `Bastion` resource in the garden cluster
+    - Mutating Webhook
+        - on creation, sets `metadata.annotations["operations.gardener.cloud/created-by"]` according to the user that created the resource
+        - when `operations.gardener.cloud/operation: keepalive` is set, calculates new `status.expirationTimestamp`
+3. `gardenlet`
+    - Watches `Bastion` resource for own seed under api group `operations.gardener.cloud` in the garden cluster
+    - Creates `Bastion` custom resource under api group `extensions.gardener.cloud/v1alpha1` in the seed cluster
+4. `GCM`:
+    - During reconcile of the `Bastion` resource:
+      - according to `spec.shootRef`, sets the `status.seedName`
+      - according to `spec.shootRef`, sets the `status.providerType`
```

Comment:
Why should GCM do this? This can be done by above mentioned admission plugin when the `Bastion` resource is created.

Reply:
@vpnachev proposed this https://github.com/gardener/gardener/pull/3802#discussion_r605505765 and I thought it was a good idea and adapted accordingly. Should I revert it?

Reply:
FMPOV, yes. I think it's just adding avoidable complexity + code if we have multiple controllers reconciling and modifying this resource. Hence, GCM should be out of the game for this particular aspect. It can be perfectly done in the admission plugin which we also need for other reasons.

Reply:
@vpnachev any objections?

Reply:
@vpnachev is out of office today. Anyways, this is an implementation detail that we can also sort out later in the actual PR adding support for it.

Reply:
Also, I think, both `seedName` and `provider` should be in the `spec` actually and by the admission plugin, like originally proposed.